### PR TITLE
Turn geo query into set of geohashes

### PR DESCRIPTION
**Pull request by agourlay**
issue context: https://github.com/qdrant/qdrant/issues/313

The goal of this PR is to turn a geo query into a set of [geohashes](https://en.wikipedia.org/wiki/Geohash) to probe later on the indexer for matching points (in a follow up PR).

To do so it introduces two main functions to deal with either circle or rectangle areas:

```rust
/// Return as-high-as-possible with maximum of `max_regions`
/// number of geo-hash guaranteed to contain the whole circle.
fn circle_hashes(circle: GeoRadius, max_regions: usize) -> Vec<GeoHash> {}

/// Return as-high-as-possible with maximum of `max_regions`
/// number of geo-hash guaranteed to contain the whole rectangle.
fn rectangle_hashes(rectangles: GeoBoundingBox, max_regions: usize) -> Vec<GeoHash> {}
```

Those functions work by estimating the `geohash` precisions possible to describe the area and select the highest precision which yields less than `max_regions`.

Notes for the reviewers:
- ~~the circle implementation relies on the rectangles implementation with additional filtering~~
- some unit tests contain comments to visualize the results in 2D

### CODE REVIEW

File: lib/segment/src/index/field_index/geo_index.rs [hunk 1]
```diff
@@ -1 +1,531 @@
+use crate::types::{GeoBoundingBox, GeoPoint, GeoRadius};
+use geo::Coordinate;
+use geohash::{decode, encode, neighbor, Direction};
+use std::ops::Range;
 
+type GeoHash = String;
```
Comment (e-ivkov) [inline lib/segment/src/index/field_index/geo_index.rs:6]:
```suggestion
struct GeoHash(String);
```
Wrapping it this way will help with type checking and maybe prevent a few bugs, as you will need to explicitly convert/wrap string into GeoHash

File: lib/segment/src/index/field_index/geo_index.rs [hunk 2]
```diff
@@ -1 +1,531 @@
...
+#[derive(Debug)]
+struct FullGeoBoundingBox {
+    north_west: GeoPoint,
+    south_west: GeoPoint,
+    south_east: GeoPoint,
+    north_east: GeoPoint,
+}
+
+impl FullGeoBoundingBox {
+    fn from_geo_bounding_box(bounding_box: &GeoBoundingBox) -> Self {
```
Comment (e-ivkov) [inline lib/segment/src/index/field_index/geo_index.rs:49]: Why not `impl From<&GeoBoundingBox> for FullGeoBoundingBox`?

File: lib/segment/src/index/field_index/geo_index.rs [hunk 3]
```diff
@@ -1 +1,531 @@
...
+
+        Self {
+            north_west,
+            south_west,
+            south_east,
+            north_east,
+        }
+    }
+
+    fn shortest_side_length_in_km(&self) -> f64 {
```
Comment (e-ivkov) [inline lib/segment/src/index/field_index/geo_index.rs:84]: Would it make sense to use https://crates.io/crates/uom ? Or do you think it is an overkill here?

File: lib/segment/src/index/field_index/geo_index.rs [hunk 4]
```diff
@@ -1 +1,531 @@
...
+    fn geohash_regions(&self) -> Vec<GeoHash> {
+        let width = self.area_width_region_count();
+        let height = self.area_height_region_count();
+
+        let mut seen: Vec<String> = Vec::new();
+        let mut top = self.north_west.clone();
+        // traverse square by columns top to bottom
+        for _i in 0..width {
+            let mut current = top.clone();
+            for _j in 0..height {
```
Comment (e-ivkov) [inline lib/segment/src/index/field_index/geo_index.rs:113]:
```suggestion
        for _ in 0..width {
            let mut current = top.clone();
            for _ in 0..height {
```
Minor stylistic suggestion

File: lib/segment/src/index/field_index/geo_index.rs (hunk 1 above)
Comment (e-ivkov) [reply lib/segment/src/index/field_index/geo_index.rs:6]: Later I checked the full logic and seems that yes, it will not help much for now.

**Comment by agourlay** [comment]
@e-ivkov I believe I have addressed all your comments :heavy_check_mark:

**Review by e-ivkov — DISMISSED** [review]

File: lib/segment/src/index/field_index/geo_index.rs [hunk 5]
```diff
@@ -1 +1,533 @@
...
+        })
+        .unwrap_or_else(|| Range { start: 1, end: 1 })
+}
+
+/// A globally-average value is usually considered to be 6,371 kilometres (3,959 mi) with a 0.3% variability (±10 km).
+/// https://en.wikipedia.org/wiki/Earth_radius.
+const EARTH_RADIUS_KM: f64 = 6371.0;
+
+/// https://rust-lang-nursery.github.io/rust-cookbook/science/mathematics/trigonometry.html#distance-between-two-points-on-the-earth
+fn distance_in_km_between_coordinates(p1: &GeoPoint, p2: &GeoPoint) -> f64 {
```
Comment (generall) [inline lib/segment/src/index/field_index/geo_index.rs:249]: We are already using https://github.com/qdrant/qdrant/blob/02a50212e516a8601829c97b7b8facf388ad7c49/lib/segment/src/payload_storage/condition_checker.rs#L6,  let's stick to one implementation for consistency

File: lib/segment/src/index/field_index/geo_index.rs [hunk 6]
```diff
@@ -1 +1,533 @@
...
+
+/// Return as-high-as-possible with maximum of `max_regions`
+/// number of geo-hash guaranteed to contain the whole circle.
+fn circle_hashes(circle: GeoRadius, max_regions: usize) -> Vec<GeoHash> {
+    let rectangle = minimum_bounding_rectangle_for_circle(&circle);
+    let inner_hashes = rectangle_hashes(rectangle, max_regions);
+    filter_hashes_within_circle(inner_hashes, circle)
+}
+
+/// filter geo hashes for which the center is within a circle
```
Comment (generall) [inline lib/segment/src/index/field_index/geo_index.rs:193]: If we only consider center of the tile - some parts of the circle could be not covered by resulting hashes. In this case it might be better to just drop this filtering and rely on exact post-filtering.

Or the alternative - to exclude only those tiles which are more distant than radius + half of the diagonal.

**Review by generall — CHANGES_REQUESTED** [review]

File: lib/segment/src/index/field_index/geo_index.rs (hunk 6 above)
Comment (generall) [reply lib/segment/src/index/field_index/geo_index.rs:182]: IMG_1324

Comment (generall) [reply lib/segment/src/index/field_index/geo_index.rs:194]: higher precision might create huge mess with trigonometry and easily take another day of work

File: lib/segment/src/index/field_index/geo_index.rs [hunk 7]
```diff
@@ -1 +1,533 @@
...
+    let shortest_side_km = full_geo_bounding_box.shortest_side_length_in_km();
+    let possible_precisions = geohash_precisions_for_distance(shortest_side_km * 1000.0);
+    // filter precision which generates less than `max_regions`.
+    let rectangle_geohash = possible_precisions
+        .into_iter()
+        .map(|p| RectangleGeoHash::compute_from_bounding_box(&rectangle, p))
+        .take_while(|rectangle_geohash| rectangle_geohash.area_region_count() <= max_regions)
+        .last();
+    match rectangle_geohash {
+        None => Vec::new(),
```
[the remark this answers is not in the document]
Comment (agourlay) [reply lib/segment/src/index/field_index/geo_index.rs:233]: > in what cases it could be empty? I guess one reason is if max_regions = 0?

In my test cases it happens rather often.

Currently the candidates geohash precisions have dimensions which are smaller than the shorter side of the rectangle.
This way we get to have several geohashes to encode the total rectangle area. This increase the granularity.

But this also makes the algorithm more restrictive and excludes the case where `max_regions = 1` by design.

We could include in the candidate geohash precisions, the smallest precision which wraps completely the area, this way we would always cover the case where `max_regions = 1` and yield at least one result. (there is a special case, this does not work if the area is bigger than a single top geohash regions).

WDYT?

[the remark this answers is not in the document]
Comment (generall) [reply lib/segment/src/index/field_index/geo_index.rs:233]: > candidates geohash precisions have dimensions which are smaller than the shorter side of the rectangle

that does not guarantee that the tile of this precision will include the whole rectangle, does it?

> But this also makes the algorithm more restrictive and excludes the case where max_regions = 1 by design.

It should include

> there is a special case, this does not work if the area is bigger than a single top geohash regions

That should result in empty geo-hash `""`

[the remark this answers is not in the document]
Comment (agourlay) [reply lib/segment/src/index/field_index/geo_index.rs]: I tried to capture your comment in the following commit https://github.com/qdrant/qdrant/pull/315/commits/5cb11072465a04cb355f3a8df636f84ff7de068d

It makes sure that no empty vector.

**Comment by agourlay** [comment]
The test case `circle_south_pole` fails and shows that the library `geohash` behaves weirdly at the limits.

Try the following case in https://www.movable-type.co.uk/scripts/geohash.html

With the geohash `5` when asking for the south neighbor we should get `g` which wraps around the planet.

But the library throws 

```
thread 'index::field_index::geo_index::tests::circle_south_pole' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidCoordinateRange(Coordinate { x: 157.5, y: -112.5 })', lib/segment/src/index/field_index/geo_index.rs:108:60
```

I am not sure if the library behaves correctly and if we need to handle this type of edge cases manually :cry:

**Comment by agourlay** [comment]
@generall all tests are now green following the protection against longitude overflow in the MBR calculation.
Ready for hopefully one last review :crossed_fingers:

**Review by generall — DISMISSED** [review]

**Review by generall — APPROVED** [review]

**Comment by agourlay** [comment]
@generall Thank you very much for the much needed refactoring :pray:
